integration/integration.go

```diff
@@ -61,49 +63,91 @@ type ServerOption interface {
 	applyServerOption(*serverOptions)
 }
 
+// TODO: What does this do?
+func WithMocks(mocks pulumi.MockResourceMonitor) ServerOption {
+	return mocksOption{mocks: mocks}
+}
+
+// WithProvider backs the server with the given concrete provider.
+func WithProvider(p provider.Provider) ServerOption {
+	return providerOption{provider: p}
+}
+
+// WithProviderF backes the server with a lazily initialized provider.
+func WithProviderF(p func(*pprovider.HostClient) p.Provider) ServerOption {
+	return providerFOption{providerF: p}
+}
+
 // serverOptions is the internal representation of the effect of
 // [ServerOption]s.
 type serverOptions struct {
+	mocks     pulumi.MockResourceMonitor
+	provider  *p.Provider
+	providerF func(*pprovider.HostClient) p.Provider
+}
+
+type mocksOption struct {
 	mocks pulumi.MockResourceMonitor
 }
 
-type serverOption func(*serverOptions)
+func (mo mocksOption) applyServerOption(opts *serverOptions) {
+	opts.mocks = mo.mocks
+}
 
-func (o serverOption) applyServerOption(opts *serverOptions) {
-	o(opts)
+type providerOption struct {
+	provider p.Provider
 }
 
-func WithMocks(mocks pulumi.MockResourceMonitor) ServerOption {
-	return serverOption(func(ro *serverOptions) {
-		ro.mocks = mocks
-	})
+func (po providerOption) applyServerOption(opts *serverOptions) {
+	opts.provider = &po.provider
 }
 
-func NewServer(pkg string, version semver.Version, provider p.Provider) Server {
-	return NewServerWithContext(context.Background(), pkg, version, provider)
+type providerFOption struct {
+	providerF func(*pprovider.HostClient) p.Provider
 }
 
-func NewServerWithContext(ctx context.Context, pkg string, version semver.Version, provider p.Provider) Server {
-	return NewServerWithOptions(ctx, pkg, version, provider)
+func (po providerFOption) applyServerOption(opts *serverOptions) {
+	opts.providerF = po.providerF
 }
 
-func NewServerWithOptions(ctx context.Context, pkg string, version semver.Version, provider p.Provider,
+// NewServer constructs a gRPC server for testing the given provider.
+//
+// Must be called with WithProvider or WithProviderF.
+func NewServer(ctx context.Context, pkg string, version semver.Version,
 	opts ...ServerOption,
-) Server {
+) (Server, error) {
 	o := &serverOptions{}
 	for _, opt := range opts {
 		opt.applyServerOption(o)
 	}
 	if o.mocks == nil {
-		o.mocks = &MockMonitor{}
+		o.mocks = &MockMonitor{} // Why?
+	}
+	if o.provider == nil && o.providerF == nil {
+		return nil, fmt.Errorf("WithProvider or WithProviderF is required")
 	}
 
 	host := newHost(ctx, o.mocks)
 
-	return &server{p.RunInfo{
+	var prov p.Provider
+	if o.provider != nil {
+		prov = *o.provider
+	}
+	if o.providerF != nil {
+		host.lazyInit()
+		hc, err := pprovider.NewHostClient(host.engineAddr)
```

Comment:
Inside `lazyInit` is a call to `NewEngineConn`, which is actually the same code as `NewHostClient`. In other words we are creating two connections but need only one.  Simply change the `host` struct to have `hostClient *pprovider.HostClient`, and then get use  `hostClient.EngineConn()` at will.